src/sc/sidechainrpc.cpp

```diff
@@ -219,6 +227,298 @@ void AddScInfoToJSON(UniValue& result)
     }
 }
 
+ScRpcCreationCmd::ScRpcCreationCmd(
+        CMutableTransaction& tx, const uint256& scid, int withdrawalEpochLength, const CBitcoinAddress& fromaddress,
+        const uint256& toaddress, const CAmount nAmount, int nMinConf, const CAmount& nFee):
+        _tx(tx), _scid(scid), _withdrawalEpochLength(withdrawalEpochLength), _fromMcAddress(fromaddress), _toScAddress(toaddress),
+        _nAmount(nAmount), _minConf(nMinConf), _fee(nFee)
+{
+    _hasFromAddress = !(fromaddress == CBitcoinAddress());
+
+    // Get dust threshold
+    CKey secret;
+    secret.MakeNewKey(true);
+    CScript scriptPubKey = GetScriptForDestination(secret.GetPubKey().GetID());
+    CTxOut out(CAmount(1), scriptPubKey);
+    _dustThreshold = out.GetDustThreshold(minRelayTxFee);
+
+    _totalInputAmount = 0;
+} 
+
+void ScRpcCreationCmd::addInputs()
+{
+    std::vector<COutput> vAvailableCoins;
+    std::vector<CmdUTXO> vInputUtxo;
+
+    static const bool fOnlyConfirmed = false;
+    static const bool fIncludeZeroValue = false;
+    static const bool fIncludeCoinBase = true;
+    static const bool fIncludeCommunityFund = true;
+
+    pwalletMain->AvailableCoins(vAvailableCoins, fOnlyConfirmed, NULL, fIncludeZeroValue, fIncludeCoinBase, fIncludeCommunityFund);
+
+    for (const auto& out: vAvailableCoins)
+    {
+        LogPrint("sc", "utxo %s depth: %d, val: %s, spendable: %s\n",
+            out.tx->GetHash().ToString(), out.nDepth, FormatMoney(out.tx->vout[out.i].nValue), out.fSpendable?"Y":"N");
+
+        if (!out.fSpendable || out.nDepth < _minConf) {
+            continue;
+        }
+
+        if (_hasFromAddress)
+        {
+            CTxDestination dest;
+            if (!ExtractDestination(out.tx->vout[out.i].scriptPubKey, dest)) {
+                continue;
+            }
+
+            if (!(CBitcoinAddress(dest) == _fromMcAddress)) {
+                continue;
+            }
+        }
+
+        CAmount nValue = out.tx->vout[out.i].nValue;
+
+        CmdUTXO utxo(out.tx->GetHash(), out.i, nValue);
+        vInputUtxo.push_back(utxo);
+    }
+
+    // sort in ascending order, so smaller utxos appear first
+    std::sort(vInputUtxo.begin(), vInputUtxo.end(), [](CmdUTXO i, CmdUTXO j) -> bool {
+        return ( std::get<2>(i) < std::get<2>(j));
+    });
+
+    CAmount totUTXOAmount = 0;
+    for (const auto & t : vInputUtxo) {
+        totUTXOAmount += std::get<2>(t);
+    }
```

Comment:
I think you can avoid this loop and perform the check totUTXOAmount < targetAmount immediately after the loop on lines 314 -327.
So you would optimistically start collecting the UTXOs you need stopping when you reach amount + target fees. You can catch the case where you have insufficient found testing for  _totalInputAmount  < targetAmount

Reply:
Ok, you are right